polyglot/piranha/src/models/source_code_unit.rs

```diff
@@ -258,43 +502,127 @@ impl SourceCodeUnit {
+      get_context(
+        self.root_node(),
+        changed_node,
+        self.code().to_string(),
+        number_of_ancestors_in_parent_scope,
+      )
+    };
+    for rule in rules {
+      for ancestor in &context() {
+        if let Some(edit) = self.get_edit(rule.clone(), rules_store, *ancestor, false) {
+          return Some(edit);
+        }
+      }
+    }
+    None
   }
 
-  pub(crate) fn path(&self) -> &PathBuf {
-    &self.path
-  }
+  /// Gets the first match for the rule in `self`
+  pub(crate) fn get_matches(
+    &self, rule: Rule, rule_store: &mut RuleStore, node: Node, recursive: bool,
+  ) -> Vec<Match> {
+    let mut output: Vec<Match> = vec![];
+    // Get all matches for the query in the given scope `node`.
+    let replace_node_tag = if rule.is_match_only_rule() || rule.is_dummy_rule() {
+      None
+    } else {
+      Some(rule.replace_node())
+    };
+    let all_query_matches = node.get_all_matches_for_query(
+      self.code().to_string(),
+      rule_store.query(&rule.query()),
+      recursive,
+      replace_node_tag,
+    );
 
-  pub(crate) fn rewrites_mut(&mut self) -> &mut Vec<Edit> {
-    &mut self.rewrites
-  }
+    // Return the first match that satisfies constraint of the rule
+    for p_match in all_query_matches {
+      let matched_node = get_node_for_range(
+        self.root_node(),
+        p_match.range().start_byte,
+        p_match.range().end_byte,
+      );
 
-  pub(crate) fn matches(&self) -> &[(String, Match)] {
-    self.matches.as_ref()
+      if matched_node.is_satisfied(&self, &rule, p_match.matches(), rule_store) {
+        trace!("Found match {:#?}", p_match);
+        output.push(p_match);
+      }
+    }
+    debug!("Matches found {}", output.len());
+    output
   }
 
-  pub(crate) fn matches_mut(&mut self) -> &mut Vec<(String, Match)> {
-    &mut self.matches
+  /// Gets the first match for the rule in `self`
+  pub(crate) fn get_edit(
+    &self, rule: Rule, rule_store: &mut RuleStore, node: Node, recursive: bool,
+  ) -> Option<Edit> {
+    // Get all matches for the query in the given scope `node`.
+
+    return self
+      .get_matches(rule.clone(), rule_store, node, recursive)
+      .first()
+      .map(|p_match| {
+        let replacement = substitute_tags(rule.replace(), p_match.matches(), false);
+        let edit = Edit::new(p_match.clone(), replacement, rule.name());
+        trace!("Rewrite found : {:#?}", edit);
+        edit
+      });
   }
 }
 
+
+pub(crate) trait SatisfiesConstraint {
+  // / Checks if the given rule satisfies the constraint of the rule, under the substitutions obtained upon matching `rule.query`
+  fn is_satisfied(&self, source_code_unit: &SourceCodeUnit, rule: &Rule, substitutions: &HashMap<String, String>,rule_store: &mut RuleStore,) -> bool ;
+}
+
+/// Checks if the given rule satisfies the constraint of the rule, under the substitutions obtained upon matching `rule.query`
+impl SatisfiesConstraint for Node<'_> {
+fn is_satisfied(
```

Comment:
Indentation is a bit weird here. Shouldn't `is_satisfied` be indented with respect to the impl block containing it? Or is this a quirk about rust formatting I am not aware of?

Reply:
I will format this. No I didnt run rust format. :|